src/dependencies/dss-direct-deposit/D3MInit.sol

```diff
@@ -0,0 +1,326 @@
+}
+
+struct D3MAavePoolConfig {
+    address king;
+    address adai;
+    address stableDebt;
+    address variableDebt;
+}
+
+struct D3MAaveRateTargetPlanConfig {
+    uint256 bar;
+    address adai;
+    address stableDebt;
+    address variableDebt;
+    address tack;
+    uint256 adaiRevision;
+}
+
+struct D3MAaveBufferPlanConfig {
+    uint256 buffer;
+    address adai;
+}
+
+struct D3MCompoundPoolConfig {
+    address king;
+    address cdai;
+    address comptroller;
+    address comp;
+}
+
+struct D3MCompoundRateTargetPlanConfig {
+    uint256 barb;
+    address cdai;
+    address tack;
+    address delegate;
+}
+
+// Init a D3M instance
+library D3MInit {
+
+    function initCore(
+        DssInstance memory dss,
+        D3MCoreInstance memory d3mCore
+    ) internal {
+        D3MHubLike hub = D3MHubLike(d3mCore.hub);
+        D3MMomLike mom = D3MMomLike(d3mCore.mom);
+
+        // Sanity checks
+        require(hub.vat() == address(dss.vat), "Hub vat mismatch");
+        require(hub.daiJoin() == address(dss.daiJoin), "Hub daiJoin mismatch");
+
+        hub.file("vow", address(dss.vow));
+        hub.file("end", address(dss.end));
+
+        mom.setAuthority(dss.chainlog.getAddress("MCD_ADM"));
+
+        dss.vat.rely(address(hub));
+
+        dss.chainlog.setAddress("DIRECT_HUB", address(hub));
+        dss.chainlog.setAddress("DIRECT_MOM", address(mom));
+    }
+
+    function initCommon(
+        DssInstance memory dss,
+        D3MInstance memory d3m,
+        D3MCommonConfig memory cfg
+    ) internal {
+        bytes32 ilk = cfg.ilk;
+        D3MHubLike hub = D3MHubLike(cfg.hub);
+        D3MOracleLike oracle = D3MOracleLike(d3m.oracle);
+
+        // Sanity checks
+        require(oracle.vat() == address(dss.vat), "Oracle vat mismatch");
+        require(oracle.ilk() == ilk, "Oracle ilk mismatch");
+
+        WardsAbstract(d3m.plan).rely(cfg.mom);
+
+        hub.file(ilk, "pool", d3m.pool);
+        hub.file(ilk, "plan", d3m.plan);
+        hub.file(ilk, "tau", cfg.tau);
+
+        oracle.file("hub", address(hub));
+
+        dss.spotter.file(ilk, "pip", address(oracle));
+        dss.spotter.file(ilk, "mat", 10 ** 27);
+        uint256 previousIlkLine;
+        if (cfg.existingIlk) {
+            (,,, previousIlkLine,) = dss.vat.ilks(ilk);
+        } else {
+            dss.vat.init(ilk);
+            dss.jug.init(ilk);
+        }
+        dss.vat.file(ilk, "line", cfg.gap);
+        dss.vat.file("Line", dss.vat.Line() + cfg.gap - previousIlkLine);
+        DssAutoLineAbstract(dss.chainlog.getAddress("MCD_IAM_AUTO_LINE")).setIlk(
+            ilk,
+            cfg.maxLine,
+            cfg.gap,
+            cfg.ttl
+        );
+        dss.spotter.poke(ilk);
+
+        GemAbstract gem = GemAbstract(ID3MPool(d3m.pool).redeemable());
+        IlkRegistryAbstract(dss.chainlog.getAddress("ILK_REGISTRY")).put(
+            ilk,
+            address(hub),
+            address(gem),
+            gem.decimals(),
+            4,
```

Comment:
`4` doesn't give much insight here, it's just a magical number. You should probably extract it into a constant like `ILK_REGISTRY_CLASS_D3M = 4`.